### PR TITLE
feat(streaming): add memory size limit to exchange

### PR DESCRIPTION
**Pull request by wangrunji0408**
Signed-off-by: Runji Wang <wangrunji0408@163.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR adds a memory size limit to exchange channels, in order to prevent potential OOM issues. This is a successor of #13775. In that PR we completely replaced row limit by memory size limit, but it has been proven to have a bad impact on performance. Therefore, in this PR, we do not abandon the row limit, but add an additional memory limit. In the implementation, we add a new `bytes` permit which will work in parallel with the other two. The default size limit is arbitrarily set to 1MB.

Besides, this PR removes the explicit threshold to send the permit back. In my opinion, ACK can be delayed but should not be delayed forever. So I changed the receiver behavior to accept all ready messages in a batch and then ACK them once. This should allow ACKs to be replied timely but not too frequently. 🤔

## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [x] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)
    The nexmark benchmark is running. https://buildkite.com/risingwave-test/nexmark-benchmark/builds/3434#018ebcc4-eaa8-4db8-9675-79f7929e3650

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

### CODE REVIEW

File: src/stream/src/executor/exchange/input.rs [hunk 1]
```diff
@@ -159,67 +156,55 @@ impl RemoteInput {
...
-                        // For barriers, always send it back immediately.
-                        Some(permits::Value::Barrier(p)) => Some(permits::Value::Barrier(p)),
-                        None => None,
-                    } {
-                        permits_tx
-                            .send(add_back_permits)
-                            .context("RemoteInput backward permits channel closed.")?;
-                    }
+        // process messages and ack permits in batch
+        let mut stream = stream.ready_chunks(8);
```
Comment (fuyufjh) [inline src/stream/src/executor/exchange/input.rs:160]: May extract `8` as a constant.

File: src/stream/src/executor/exchange/input.rs [hunk 2]
```diff
@@ -159,67 +156,55 @@ impl RemoteInput {
...
+
+                let permits = permits.unwrap();
+                ack_permits.records += permits.records;
+                ack_permits.bytes += permits.bytes;
+                ack_permits.barriers += permits.barriers;
+            }
+
+            metrics
+                .exchange_frag_recv_size
+                .with_label_values(&[&up_fragment_id, &down_fragment_id])
```
Comment (fuyufjh) [inline src/stream/src/executor/exchange/input.rs:185]: (off-topic) Might need to be refactored #12805

**Review by fuyufjh — APPROVED** [review]
LGTM. Please do some benchmark & longevity before merging

**Review by BugenZhao — APPROVED** [review]
> So I changed the receiver behavior to accept all ready messages in a batch and then ACK them once. This should allow ACKs to be replied timely but not too frequently. 🤔

I'm not sure if it works as well as before, but it sounds promising. Let's give it a try.

**Comment by wangrunji0408** [comment]
Performance degraded in nexmark q0, q5(not shown, about -20%), q15, q17, q20. 😕
improved: q8.

截屏2024-04-24 10 38 05

**Pull request closed without merging**